src/ui/public/chrome/api/nav.js

```diff
@@ -40,25 +40,72 @@ export default function (chrome, internals) {
   }
 
   function refreshLastUrl(link) {
-    link.lastSubUrl = internals.appUrlStore.getItem(lastSubUrlKey(link));
+    link.lastSubUrl = internals.appUrlStore.getItem(lastSubUrlKey(link)) || link.lastSubUrl || link.url;
+  }
+
+  function getAppId(url) {
+    const pathname = parse(url).pathname;
+    const pathnameWithoutBasepath = pathname.slice(chrome.getBasePath().length);
+    const match = String(pathnameWithoutBasepath).match(/^\/app\/([^\/]+)(?:\/|\?|#|$)/);
```

Comment:
Since String.prototype.slice() returns a string, I think the String() call here is unnecessarily defensive.  We'd actually probably want it to throw in that situation since it would be very unexpected.